spotlight/factorization/implicit.py

```diff
@@ -236,7 +247,106 @@ def fit(self, interactions, verbose=False):
 
                 self._optimizer.zero_grad()
 
-                loss = self._loss_func(positive_prediction, negative_prediction)
+                loss = self._loss_func(
+                    positive_prediction,
+                    negative_prediction,
+                )
+                epoch_loss += loss.item()
+
+                loss.backward()
+                self._optimizer.step()
+
+            epoch_loss /= minibatch_num + 1
+
+            if verbose:
+                print('Epoch {}: loss {}'.format(epoch_num, epoch_loss))
+
+            if np.isnan(epoch_loss) or epoch_loss == 0.0:
+                raise ValueError('Degenerate epoch loss: {}'
+                                 .format(epoch_loss))
+
+    def _fit_weighted(self, interactions, verbose=False):
```

Comment:
I'm still not convinced about having a separate `_fit_weighted` function, even if it is internal. I think it introduces too much code duplication that will have to be kept in sync.

What about modifying `minibatch` to look roughly like this:
```python
def minibatch(*tensors, **kwargs):

    batch_size = kwargs.get('batch_size', 128)

    if len(tensors) == 1:
        tensor = tensors[0]
        for i in range(0, len(tensor), batch_size):
            yield tensor[i:i + batch_size]
    else:
        for i in range(0, len(tensors[0]), batch_size):
            yield tuple(x[i:i + batch_size] if x is not None else None for x in tensors)
```
This way, it emits tensor slices if an argument is a tensor (as before), but also emits `None` in the tuple if an argument is `None`.